test/string.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <stddef.h>
+#include <string.h>
+#include <unity.h>
+
+static void
+test_strcmp(void)
+{
+	/* Test with long string. */
+	TEST_ASSERT_EQUAL(0, strcmp("Here is a string", "Here is a string"));
+
+	/* Test with medium string. */
+	TEST_ASSERT_EQUAL(0, strcmp("Hello World", "Hello World"));
+
+	/* Test with short string. */
+	TEST_ASSERT_EQUAL(0, strcmp("A", "A"));
+
+	/* Test with numbers in string. */
+	TEST_ASSERT_EQUAL(0, strcmp("He11o W0rld", "He11o W0rld"));
+
+	/* Test with special characters. */
+	TEST_ASSERT_EQUAL(0, strcmp("Hello_World!", "Hello_World!"));
+
+	/* Test string with only a space. */
+	TEST_ASSERT_EQUAL(0, strcmp(" ", " "));
+
+	/* Test an empty string. */
+	TEST_ASSERT_EQUAL(0, strcmp("", ""));
+
+	/* Test two different strings. */
+	int result = strcmp("Here is a string", "Here is a different string");
+	TEST_ASSERT(result == 15);
+
+	int result2 = strcmp("Here is a different string", "Here is a string");
+	TEST_ASSERT(result2 == -15);
+}
+
+static void
+test_strlen(void)
+{
+	/* Test with long string. */
+	TEST_ASSERT_EQUAL(59, strlen("Here is a long string. "
+	                             "This goes above thirty-two charcters"));
+
+	/* Test with medium string. */
+	TEST_ASSERT_EQUAL(23, strlen("Here is a medium string"));
+
+	/* Test with short string. */
+	TEST_ASSERT_EQUAL(6, strlen("String"));
+
+	/* Test with numbers in string. */
+	TEST_ASSERT_EQUAL(7, strlen("String7"));
+
+	/* Test with special characters. */
+	TEST_ASSERT_EQUAL(8, strlen("Str-ing!"));
+
+	/* Test with empty string. */
+	TEST_ASSERT_EQUAL(0, strlen(""));
+}
+
+static void
+test_memcpy(void)
+{
+	/* Test a valid string.*/
+	char src[30] = "This is a test of Memcpy";
+	char dest[30];
+
+	memcpy(dest, src, (sizeof(dest)));
+	TEST_ASSERT_EQUAL_STRING(src, dest);
+
+	/* Test a valid string with special characters. */
+	char src2[50] = "This_is_a_test_of_Memcpy!";
+	char dest2[50];
+	memcpy(dest2, src2, sizeof(dest2));
+	TEST_ASSERT_EQUAL_STRING(src2, dest2);
+
+	/* Test with a uint8_t array */
+	uint8_t src3[] = { 0x00, 0x01 };
+	uint8_t dest3[10];
```

Comment:
The source and destination should actually be the same size in bytes (reading past the end is Undefined Behavior).

Reply:
what do you mean?

Reply:
done